### PR TITLE
#163383176 Send verification email after user registration

### PR DESCRIPTION
**Pull request by codjoero**
#### What does this PR do?

- add email notification method to registrationAPIView
- add accountActivateAPIView
- implement token decoding
- add url for accountActivateAPIView
- write tests for the Registration api view class
- test more functionality to raise coverage

#### Description of Task to be completed?

Users should receive an email containing an activation link when they enter valid user parameters in an attempt to signup.
Users' accounts should be activated when they click on the activation link they received in their email

#### How should this be manually tested?

Visit this [PR's app](https://ah-backend-summer-stagin-pr-21.herokuapp.com/)

#### Any background context you want to provide?

The App was less secure without authentication of emails provided by users

#### What are the relevant pivotal tracker stories?

[#163383176](https://www.pivotaltracker.com/story/show/163383176)

### CODE REVIEW

**Review by fahdjamy — APPROVED** [review]
Thanks for maintaining the coverage.
LGTM now.

**Review by oma0256 — APPROVED** [review]
@codjoero looks good to me

**Review by missvicki — APPROVED** [review]

File: authors/apps/authentication/tests/test_registration.py [hunk 1]
```diff
@@ -16,13 +20,86 @@ def test_user_register_successful(self):
...
+        """
+        Test user can not access activateion view with an expired token.
+        """
+        url = expired_link
+        response = self.client.get(url, content_type='application/json')
+        self.assertEqual(
+                    response.data['err_msg'], "Token expired")
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_cannot_activate_with_invalid_link(self):
```
Comment (malep2007) [inline authors/apps/authentication/tests/test_registration.py]: Hello @codjoero, I think you may have to start breaking these tests to create `unittests` and `integration` tests. The tests you've developed thus far do not show a clear distinction between the two. I'd recommend further reading of `Unittests vs integration` tests. You can start [here.](https://www.softwaretestingclass.com/what-is-difference-between-unit-testing-and-integration-testing/)

Comment (codjoero) [reply authors/apps/authentication/tests/test_registration.py]: Hello @malep2007, thanks for the feedback and resource. I think I understand the difference now. Would you like to see the different types in their separate folders? Thanks.

File: authors/apps/authentication/tests/base_class.py [hunk 2]
```diff
@@ -22,16 +22,22 @@ def setUp(self):
...
 
     def register_and_login_test_user(self):
         self.register_test_user()
-        # force authentication for now, since the login system is not implemented yet
-        test_user = User.objects.get(email=valid_login_data['user']['email'])
-        self.client.force_authenticate(user=test_user)
-        # we'll use this when the token based auth system is ready
-        # self.client.credentials(HTTPS_AUTHORIZATION="Bearer {}".format(
-        #     self.client.post(self.url_login, data=valid_login_data, format='json').data["token"]))
+        activation_link = (mail.outbox[0].body.split('\n')).pop(1)
```
Comment (karuhanga) [inline authors/apps/authentication/tests/base_class.py:40]: I think this part of the code needs a bit more of comments.
It is not clear what you are trying to do.

File: authors/apps/authentication/tests/test_login.py [hunk 3]
```diff
@@ -11,11 +12,17 @@
 class LoginTest(BaseTest):
     def setUp(self):
         super().setUp()
-        self.register_test_user()
+        self.resp = self.register_test_user()
 
     def test_login_successful(self):
```
Comment (karuhanga) [inline authors/apps/authentication/tests/test_login.py:14]: I'll add on to what @malep2007 noted and mention separation of concerns. This test in particular seems to combine testing of two independent functionality; `activation` and `login`.

**Review by karuhanga — CHANGES_REQUESTED** [review]
Good work @codjoero. Most of the code for this PR is clear and easy to follow.
A few observations though;
- The `activated_user` and `register_and_login_test_user` methods in `base_class.py` seem to achieve the same purpose. Some clarification please.

File: authors/apps/authentication/tests/base_class.py (hunk 2 above)
Comment (codjoero) [reply authors/apps/authentication/tests/base_class.py:40]: Thanks for the feedback @Karuhanga I will go ahead and add a descriptive doc-string. `outbox` is a great tool to use when testing outgoing emails from Django's `mail` class. Kindly look at [this](https://docs.djangoproject.com/en/2.1/topics/testing/tools/#email-services) documentation.

File: authors/apps/authentication/tests/test_login.py (hunk 3 above)
Comment (codjoero) [reply authors/apps/authentication/tests/test_login.py:14]: Thanks for that catch @Karuhanga, `refactoring...`

File: authors/apps/authentication/tests/test_registration.py [hunk 4]
```diff
@@ -16,13 +20,86 @@ def test_user_register_successful(self):
...
                          valid_register_data["user"]["username"])
 
+    def test_activation_email_is_sent_on_valid_registration(self):
+        """
+        Test activation email is sent to user email upon
+        succesful registration.
+        """
+        self.register_test_user()
+        self.assertEqual(
+                    "Activation for your Author's Heaven account",
```
Comment (missvicki) [inline authors/apps/authentication/tests/test_registration.py]: @codjoero Typing error on Author's Haven, It's Haven not Heaven

File: authors/apps/authentication/urls.py [hunk 5]
```diff
@@ -12,4 +13,5 @@
     path('users/login/twitter/', TwitterAuthAPIView.as_view(), name='twitter-auth'),
     path('users/login/google/', GoogleAuthAPIView.as_view(), name='google-auth'),
     path('users/login/facebook/', FacebookAuthAPIView.as_view(), name='facebook-auth'),
+    path('users/<token>/', AccountActivateAPIView.as_view(), name='activate'),
```
Comment (missvicki) [inline authors/apps/authentication/urls.py:16]: @codjoero I feel that writing the namespace here as activate-account would be more descriptive than just activate. What do you think?

File: authors/apps/authentication/views.py [hunk 6]
```diff
@@ -28,17 +34,99 @@ class RegistrationAPIView(GenericAPIView):
     serializer_class = RegistrationSerializer
 
     def post(self, request):
+        """
```
Comment (missvicki) [inline authors/apps/authentication/views.py:37]: Great doc-string description, I have learned something from this and I will be taking it up as well in the next few implementations. Thank you.

Comment (codjoero) [reply authors/apps/authentication/views.py:37]: We learn and progress, human evolution.

File: authors/apps/authentication/urls.py (hunk 5 above)
Comment (codjoero) [reply authors/apps/authentication/urls.py:16]: @missvicki , great suggestion, adding...

File: authors/apps/authentication/tests/test_registration.py (hunk 4 above)
Comment (codjoero) [reply authors/apps/authentication/tests/test_registration.py]: resolved, thanks

**Comment by malep2007** [comment]
@codjoero ,please fix merge conflicts

**Pull request merged**